Unsupported type: array(bigint) in TupleDomainFilterUtils

Comparing array column with literal array triggers an error in TupleDomainFilterUtils:

```
CREATE TABLE lineitem_ex AS
SELECT orderkey, ARRAY[orderkey, partkey, suppkey] AS keys FROM lineitem;

SELECT orderkey FROM lineitem_ex WHERE keys = ARRAY[1, 2, 3];
```

```
Caused by: java.lang.UnsupportedOperationException: Unsupported type: array(bigint)
	at com.facebook.presto.orc.TupleDomainFilterUtils.createRangeFilter(TupleDomainFilterUtils.java:173)
	at com.facebook.presto.orc.TupleDomainFilterUtils.toFilter(TupleDomainFilterUtils.java:76)
	at com.google.common.collect.Maps$9.transformEntry(Maps.java:1955)
	at com.google.common.collect.Maps$12.getValue(Maps.java:1996)
	at com.facebook.presto.hive.orc.OrcSelectivePageSourceFactory.toTupleDomainFilters(OrcSelectivePageSourceFactory.java:319)
	at com.facebook.presto.hive.orc.OrcSelectivePageSourceFactory.createOrcPageSource(OrcSelectivePageSourceFactory.java:249)
	... 17 more
```

CC: @prestodb/aria

--- a/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java
+++ b/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java
@@ -1725,7 +1725,7 @@ public class HiveMetadata
         //  - range filters that apply to subfields,
         //  - the rest
         ExtractionResult<Subfield> decomposedFilter = rowExpressionService.getDomainTranslator()
-                .fromPredicate(session, filter, new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session));
+                .fromPredicate(session, filter, new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session).toColumnExtractor());
 
         Map<String, ColumnHandle> columnHandles = getColumnHandles(session, tableHandle);
         TupleDomain<ColumnHandle> entireColumnDomain = decomposedFilter.getTupleDomain()

--- a/presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java
+++ b/presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java
@@ -23,7 +23,10 @@ import com.facebook.presto.spi.relation.ExpressionOptimizer;
 import com.facebook.presto.spi.relation.RowExpression;
 import com.facebook.presto.spi.relation.SpecialFormExpression;
 import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.spi.type.ArrayType;
+import com.facebook.presto.spi.type.MapType;
 import com.facebook.presto.spi.type.RowType;
+import com.facebook.presto.spi.type.Type;
 import io.airlift.slice.Slice;
 
 import java.util.ArrayList;
@@ -37,7 +40,6 @@ import static java.util.Collections.unmodifiableList;
 import static java.util.Objects.requireNonNull;
 
 public final class SubfieldExtractor
-        implements DomainTranslator.ColumnExtractor<Subfield>
 {
     private final StandardFunctionResolution functionResolution;
     private final ExpressionOptimizer expressionOptimizer;
@@ -53,7 +55,21 @@ public final class SubfieldExtractor
         this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
     }
 
-    @Override
+    public DomainTranslator.ColumnExtractor toColumnExtractor()
+    {
+        return (expression, domain) -> {
+            // Only allow null checks on complex types
+            Type type = expression.getType();
+            if (type instanceof ArrayType || type instanceof MapType || type instanceof RowType) {
+                if (!domain.isOnlyNull() && !(domain.getValues().isAll() && !domain.isNullAllowed())) {
+                    return Optional.empty();
+                }
+            }
+
+            return extract(expression);
+        };
+    }
+
     public Optional<Subfield> extract(RowExpression expression)
     {
         return toSubfield(expression, functionResolution, expressionOptimizer, connectorSession);

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java
@@ -20,8 +20,10 @@ import com.facebook.presto.spi.function.FunctionHandle;
 import com.facebook.presto.spi.function.OperatorType;
 import com.facebook.presto.spi.predicate.Domain;
 import com.facebook.presto.spi.predicate.Range;
+import com.facebook.presto.spi.predicate.TupleDomain;
 import com.facebook.presto.spi.predicate.ValueSet;
 import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.DomainTranslator;
 import com.facebook.presto.spi.relation.DomainTranslator.ExtractionResult;
 import com.facebook.presto.spi.relation.RowExpression;
 import com.facebook.presto.spi.relation.SpecialFormExpression;
@@ -42,13 +44,17 @@ import java.util.List;
 import java.util.Map;
 
 import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.block.BlockAssertions.createArrayBigintBlock;
+import static com.facebook.presto.block.BlockAssertions.createMapBlock;
 import static com.facebook.presto.hive.HiveTestUtils.mapType;
 import static com.facebook.presto.metadata.MetadataManager.createTestMetadataManager;
 import static com.facebook.presto.spi.function.OperatorType.SUBSCRIPT;
 import static com.facebook.presto.spi.predicate.TupleDomain.withColumnDomains;
+import static com.facebook.presto.spi.relation.ConstantExpression.createConstantExpression;
 import static com.facebook.presto.spi.relation.LogicalRowExpressions.TRUE_CONSTANT;
 import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.DEREFERENCE;
 import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IN;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IS_NULL;
 import static com.facebook.presto.spi.type.BigintType.BIGINT;
 import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
 import static com.facebook.presto.spi.type.IntegerType.INTEGER;
@@ -75,7 +81,7 @@ public class TestDomainTranslator
 
     private Metadata metadata;
     private RowExpressionDomainTranslator domainTranslator;
-    private SubfieldExtractor columnExtractor;
+    private DomainTranslator.ColumnExtractor<Subfield> columnExtractor;
 
     @BeforeClass
     public void setup()
@@ -85,7 +91,7 @@ public class TestDomainTranslator
         columnExtractor = new SubfieldExtractor(
                 new FunctionResolution(metadata.getFunctionManager()),
                 (rowExpression, level, session) -> rowExpression,
-                TestingSession.SESSION);
+                TestingSession.SESSION).toColumnExtractor();
     }
 
     @Test
@@ -117,12 +123,32 @@ public class TestDomainTranslator
 
             assertPredicateTranslates(bigintIn(expression, ImmutableList.of(1L, 2L)), subfield, Domain.create(ValueSet.ofRanges(Range.equal(BIGINT, 1L), Range.equal(BIGINT, 2L)), false));
         }
+
+        Type arrayType = C_BIGINT_ARRAY.getType();
+        assertPredicateTranslates(isNull(C_BIGINT_ARRAY), C_BIGINT_ARRAY.getName(), Domain.create(ValueSet.none(arrayType), true));
+        assertPredicateTranslates(not(isNull(C_BIGINT_ARRAY)), C_BIGINT_ARRAY.getName(), Domain.create(ValueSet.all(arrayType), false));
+        assertPredicateDoesNotTranslate(equal(C_BIGINT_ARRAY, createConstantExpression(createArrayBigintBlock(ImmutableList.of(ImmutableList.of(1L, 2L, 3L))), arrayType)));
+
+        MapType mapType = (MapType) C_BIGINT_TO_BIGINT_MAP.getType();
+        assertPredicateTranslates(isNull(C_BIGINT_TO_BIGINT_MAP), C_BIGINT_TO_BIGINT_MAP.getName(), Domain.create(ValueSet.none(mapType), true));
+        assertPredicateTranslates(not(isNull(C_BIGINT_TO_BIGINT_MAP)), C_BIGINT_TO_BIGINT_MAP.getName(), Domain.create(ValueSet.all(mapType), false));
+        assertPredicateDoesNotTranslate(equal(C_BIGINT_TO_BIGINT_MAP, createConstantExpression(createMapBlock(mapType, ImmutableMap.of(1, 100)), mapType)));
     }
 
     private RowExpression dereference(RowExpression base, int field)
     {
         Type fieldType = base.getType().getTypeParameters().get(field);
         return specialForm(DEREFERENCE, fieldType, ImmutableList.of(base, new ConstantExpression((long) field, INTEGER)));
+    }
+
+    private RowExpression isNull(RowExpression expression)
+    {
+        return specialForm(IS_NULL, BOOLEAN, expression);
+    }
+
+    private RowExpression not(RowExpression expression)
+    {
+        return call("not", new FunctionResolution(metadata.getFunctionManager()).notFunction(), BOOLEAN, expression);
     }
 
     private RowExpression arraySubscript(RowExpression arrayExpression, int index)
@@ -154,6 +180,13 @@ public class TestDomainTranslator
         ExtractionResult<Subfield> result = domainTranslator.fromPredicate(TEST_SESSION.toConnectorSession(), predicate, columnExtractor);
         assertEquals(result.getRemainingExpression(), TRUE_CONSTANT);
         assertEquals(result.getTupleDomain(), withColumnDomains(ImmutableMap.of(new Subfield(subfield), domain)));
+    }
+
+    private void assertPredicateDoesNotTranslate(RowExpression predicate)
+    {
+        ExtractionResult<Subfield> result = domainTranslator.fromPredicate(TEST_SESSION.toConnectorSession(), predicate, columnExtractor);
+        assertEquals(result.getTupleDomain(), TupleDomain.all());
+        assertEquals(result.getRemainingExpression(), predicate);
     }
 
     private RowExpression greaterThan(RowExpression left, RowExpression right)

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java
@@ -42,6 +42,7 @@ import java.util.List;
 import java.util.Map;
 
 import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.hive.HiveTestUtils.mapType;
 import static com.facebook.presto.metadata.MetadataManager.createTestMetadataManager;
 import static com.facebook.presto.spi.function.OperatorType.SUBSCRIPT;
 import static com.facebook.presto.spi.predicate.TupleDomain.withColumnDomains;
@@ -56,7 +57,6 @@ import static com.facebook.presto.sql.analyzer.TypeSignatureProvider.fromTypes;
 import static com.facebook.presto.sql.relational.Expressions.call;
 import static com.facebook.presto.sql.relational.Expressions.constant;
 import static com.facebook.presto.sql.relational.Expressions.specialForm;
-import static com.facebook.presto.util.Reflection.methodHandle;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static org.testng.Assert.assertEquals;
 
@@ -142,22 +142,6 @@ public class TestDomainTranslator
                 operator(SUBSCRIPT, mapType(mapType.getKeyType(), mapType.getValueType()), mapType.getKeyType()),
                 mapType.getValueType(),
                 ImmutableList.of(mapExpression, keyExpression));
-    }
-
-    private static MapType mapType(Type keyType, Type valueType)
-    {
-        return new MapType(
-                keyType,
-                valueType,
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"),
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"),
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"),
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"));
-    }
-
-    public static void throwUnsupportedOperationException()
-    {
-        throw new UnsupportedOperationException();
     }
 
     private FunctionHandle operator(OperatorType operatorType, Type... types)

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java
@@ -293,6 +293,9 @@ public class TestHivePushdownFilterQueries
             // filter-only map
             assertQueryUsingH2Cte("SELECT linenumber FROM test_maps WHERE map_keys IS NOT NULL", rewriter);
 
+            // equality filter
+            assertQuery("SELECT orderkey FROM test_maps WHERE map_flags = MAP(ARRAY[1, 2], ARRAY[true, true])", "SELECT orderkey FROM lineitem WHERE orderkey % 17 <> 0 AND shipmode = 'AIR' AND returnflag = 'R'");
+
             assertQueryFails("SELECT map_keys[5] FROM test_maps WHERE map_keys[1] % 2 = 0", "Key not present in map: 5");
             assertQueryFails("SELECT map_keys[5] FROM test_maps WHERE map_keys[4] % 2 = 0", "Key not present in map: 4");
         }
@@ -376,6 +379,9 @@ public class TestHivePushdownFilterQueries
         assertFilterProject("nested_keys[3] IS NOT NULL AND nested_keys[1][2] > 10", "keys, flags");
         assertFilterProject("nested_keys IS NOT NULL AND nested_keys[3] IS NOT NULL AND nested_keys[1][1] > 0", "keys");
         assertFilterProject("nested_keys IS NOT NULL AND nested_keys[3] IS NULL AND nested_keys[1][1] > 0", "keys");
+
+        // equality filter
+        assertQuery("SELECT orderkey FROM lineitem_ex WHERE keys = ARRAY[1, 22, 48]", "SELECT orderkey FROM lineitem WHERE orderkey = 1 AND partkey = 22 AND suppkey = 48");
 
         assertFilterProjectFails("keys[5] > 0", "orderkey", "Array subscript out of bounds");
         assertFilterProjectFails("nested_keys[5][1] > 0", "orderkey", "Array subscript out of bounds");

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestSubfieldExtractor.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestSubfieldExtractor.java
@@ -33,6 +33,7 @@ import org.testng.annotations.Test;
 
 import java.util.Optional;
 
+import static com.facebook.presto.hive.HiveTestUtils.mapType;
 import static com.facebook.presto.metadata.MetadataManager.createTestMetadataManager;
 import static com.facebook.presto.spi.function.OperatorType.SUBSCRIPT;
 import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.DEREFERENCE;
@@ -44,7 +45,6 @@ import static com.facebook.presto.sql.relational.Expressions.call;
 import static com.facebook.presto.sql.relational.Expressions.constant;
 import static com.facebook.presto.sql.relational.Expressions.specialForm;
 import static com.facebook.presto.testing.assertions.Assert.assertEquals;
-import static com.facebook.presto.util.Reflection.methodHandle;
 
 public class TestSubfieldExtractor
 {
@@ -116,22 +116,6 @@ public class TestSubfieldExtractor
                 operator(SUBSCRIPT, mapType(mapType.getKeyType(), mapType.getValueType()), mapType.getKeyType()),
                 mapType.getValueType(),
                 ImmutableList.of(mapExpression, keyExpression));
-    }
-
-    private static MapType mapType(Type keyType, Type valueType)
-    {
-        return new MapType(
-                keyType,
-                valueType,
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"),
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"),
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"),
-                methodHandle(TestDomainTranslator.class, "throwUnsupportedOperationException"));
-    }
-
-    public static void throwUnsupportedOperationException()
-    {
-        throw new UnsupportedOperationException();
     }
 
     private FunctionHandle operator(OperatorType operatorType, Type... types)

--- a/presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionDomainTranslator.java
+++ b/presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionDomainTranslator.java
@@ -348,12 +348,12 @@ public final class RowExpressionDomainTranslator
                 }
                 case IS_NULL: {
                     RowExpression value = node.getArguments().get(0);
-                    Optional<T> column = columnExtractor.extract(value);
+                    Domain domain = complementIfNecessary(Domain.onlyNull(value.getType()), complement);
+                    Optional<T> column = columnExtractor.extract(value, domain);
                     if (!column.isPresent()) {
                         return visitRowExpression(node, complement);
                     }
 
-                    Domain domain = complementIfNecessary(Domain.onlyNull(value.getType()), complement);
                     return new ExtractionResult<>(TupleDomain.withColumnDomains(ImmutableMap.of(column.get(), domain)), TRUE_CONSTANT);
                 }
                 default:
@@ -409,13 +409,17 @@ public final class RowExpressionDomainTranslator
                 NormalizedSimpleComparison normalized = optionalNormalized.get();
 
                 RowExpression expression = normalized.getExpression();
-                Optional<T> column = columnExtractor.extract(expression);
+                NullableValue value = normalized.getValue();
+                Domain domain = createComparisonDomain(normalized.getComparisonOperator(), value.getType(), value.getValue(), complement);
+                Optional<T> column = columnExtractor.extract(expression, domain);
                 if (column.isPresent()) {
-                    NullableValue value = normalized.getValue();
-                    Type type = value.getType(); // common type for symbol and value
-                    return createComparisonExtractionResult(normalized.getComparisonOperator(), column.get(), type, value.getValue(), complement);
+                    if (domain.isNone()) {
+                        return new ExtractionResult<>(TupleDomain.none(), TRUE_CONSTANT);
+                    }
+                    return new ExtractionResult<>(TupleDomain.withColumnDomains(ImmutableMap.of(column.get(), domain)), TRUE_CONSTANT);
                 }
-                else if (expression instanceof CallExpression && resolution.isCastFunction(((CallExpression) expression).getFunctionHandle())) {
+
+                if (expression instanceof CallExpression && resolution.isCastFunction(((CallExpression) expression).getFunctionHandle())) {
                     CallExpression castExpression = (CallExpression) expression;
                     if (!isImplicitCoercion(castExpression)) {
                         //
@@ -669,7 +673,7 @@ public final class RowExpressionDomainTranslator
             return Optional.of(new NormalizedSimpleComparison(expression, comparisonOperator, value));
         }
 
-        private static <T> ExtractionResult<T> createComparisonExtractionResult(OperatorType comparisonOperator, T column, Type type, @Nullable Object value, boolean complement)
+        private static Domain createComparisonDomain(OperatorType comparisonOperator, Type type, @Nullable Object value, boolean complement)
         {
             if (value == null) {
                 switch (comparisonOperator) {
@@ -679,13 +683,10 @@ public final class RowExpressionDomainTranslator
                     case LESS_THAN:
                     case LESS_THAN_OR_EQUAL:
                     case NOT_EQUAL:
-                        return new ExtractionResult<>(TupleDomain.none(), TRUE_CONSTANT);
+                        return Domain.none(type);
 
                     case IS_DISTINCT_FROM:
-                        Domain domain = complementIfNecessary(Domain.notNull(type), complement);
-                        return new ExtractionResult<>(
-                                TupleDomain.withColumnDomains(ImmutableMap.of(column, domain)),
-                                TRUE_CONSTANT);
+                        return complementIfNecessary(Domain.notNull(type), complement);
 
                     default:
                         throw new AssertionError("Unhandled operator: " + comparisonOperator);
@@ -703,7 +704,7 @@ public final class RowExpressionDomainTranslator
                 throw new AssertionError("Type cannot be used in a comparison expression (should have been caught in analysis): " + type);
             }
 
-            return new ExtractionResult<>(TupleDomain.withColumnDomains(ImmutableMap.of(column, domain)), TRUE_CONSTANT);
+            return domain;
         }
 
         private static OperatorType flip(OperatorType operatorType)

--- a/presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java
+++ b/presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java
@@ -19,6 +19,7 @@ import com.facebook.presto.spi.block.DictionaryBlock;
 import com.facebook.presto.spi.block.RunLengthEncodedBlock;
 import com.facebook.presto.spi.type.ArrayType;
 import com.facebook.presto.spi.type.DecimalType;
+import com.facebook.presto.spi.type.MapType;
 import com.facebook.presto.spi.type.Type;
 import io.airlift.slice.Slice;
 
@@ -28,6 +29,7 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 
 import static com.facebook.presto.spi.type.BigintType.BIGINT;
 import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
@@ -43,6 +45,7 @@ import static com.facebook.presto.spi.type.TimestampWithTimeZoneType.TIMESTAMP_W
 import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
 import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
 import static com.facebook.presto.testing.TestingConnectorSession.SESSION;
+import static com.facebook.presto.util.StructuralTestUtil.appendToBlockBuilder;
 import static com.google.common.base.Preconditions.checkArgument;
 import static io.airlift.slice.Slices.wrappedBuffer;
 import static java.lang.Float.floatToRawIntBits;
@@ -176,6 +179,18 @@ public final class BlockAssertions
         }
 
         return builder.build();
+    }
+
+    public static <K, V> Block createMapBlock(MapType type, Map<K, V> map)
+    {
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, map.size());
+        for (Map.Entry<K, V> entry : map.entrySet()) {
+            BlockBuilder entryBuilder = blockBuilder.beginBlockEntry();
+            appendToBlockBuilder(BIGINT, entry.getKey(), entryBuilder);
+            appendToBlockBuilder(BIGINT, entry.getValue(), entryBuilder);
+            blockBuilder.closeEntry();
+        }
+        return blockBuilder.build();
     }
 
     public static Block createBooleansBlock(Boolean... values)

--- a/presto-spi/src/main/java/com/facebook/presto/spi/relation/DomainTranslator.java
+++ b/presto-spi/src/main/java/com/facebook/presto/spi/relation/DomainTranslator.java
@@ -14,6 +14,7 @@
 package com.facebook.presto.spi.relation;
 
 import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.predicate.Domain;
 import com.facebook.presto.spi.predicate.TupleDomain;
 
 import java.util.Optional;
@@ -24,7 +25,13 @@ public interface DomainTranslator
 {
     interface ColumnExtractor<T>
     {
-        Optional<T> extract(RowExpression expression);
+        /**
+         * Given an expression and values domain, determine whether the expression qualifies as a
+         * "column" and return its desired representation.
+         *
+         * Return Optional.empty() if expression doesn't qualify.
+         */
+        Optional<T> extract(RowExpression expression, Domain domain);
     }
 
     RowExpression toPredicate(TupleDomain<VariableReferenceExpression> tupleDomain);
@@ -59,7 +66,7 @@ public interface DomainTranslator
         }
     }
 
-    ColumnExtractor<VariableReferenceExpression> BASIC_COLUMN_EXTRACTOR = expression -> {
+    ColumnExtractor<VariableReferenceExpression> BASIC_COLUMN_EXTRACTOR = (expression, domain) -> {
         if (expression instanceof VariableReferenceExpression) {
             return Optional.of((VariableReferenceExpression) expression);
         }

